src/rdb/rdb_raft.c

```diff
@@ -227,152 +240,129 @@ rdb_lc_store_replicas(daos_handle_t lc, uint64_t index,
 			     2 /* n */, keys, vals);
 }
 
+static int
+rdb_raft_load_replicas(daos_handle_t lc, uint64_t index, d_rank_list_t **replicas)
+{
+	d_iov_t		value;
+	uint8_t		nreplicas;
+	d_rank_list_t  *r;
+	int		rc;
+
+	d_iov_set(&value, &nreplicas, sizeof(nreplicas));
+	rc = rdb_lc_lookup(lc, index, RDB_LC_ATTRS, &rdb_lc_nreplicas, &value);
+	if (rc == -DER_NONEXIST) {
+		D_DEBUG(DB_MD, "no replicas in "DF_U64"\n", index);
+		rc = 0;
+		nreplicas = 0;
+	} else if (rc != 0) {
+		return rc;
+	}
+
+	r = daos_rank_list_alloc(nreplicas);
+	if (r == NULL)
+		return -DER_NOMEM;
+
+	if (nreplicas > 0) {
+		d_iov_set(&value, r->rl_ranks, sizeof(*r->rl_ranks) * nreplicas);
+		rc = rdb_lc_lookup(lc, index, RDB_LC_ATTRS, &rdb_lc_replicas, &value);
+		if (rc != 0) {
+			d_rank_list_free(r);
+			return rc;
+		}
+	}
+
+	*replicas = r;
+	return 0;
+}
+
 /* Caller must hold d_raft_mutex. */
 static int
 rdb_raft_add_node(struct rdb *db, d_rank_t rank)
 {
 	struct rdb_raft_node	*dnode;
 	raft_node_t		*node;
-	d_rank_t		 self;
 	int			 rc = 0;
 
-	D_ALLOC_PTR(dnode);
+	/*
+	 * Note that we are unable to handle failures from this allocation at
+	 * the moment. See also rdb_raft_cb_notify_membership_event and
+	 * rdb_raft_load_snapshot.
+	 */
+	dnode = calloc(1, sizeof(*dnode));
```

Comment:
so you changed this to calloc instead of macro because we inject failure through the macro?

Reply:
@gnailzenh, yes. There's a comment in rdb_raft_cb_notify_membership_event that states this reason explicitly.